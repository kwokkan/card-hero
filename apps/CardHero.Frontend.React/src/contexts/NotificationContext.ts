import React from "react";
import { INotificationItem } from "../types/INotificationItem";

export interface INotificationContextProps {
    notifications: INotificationItem[];
    addNotification: (notification: INotificationItem) => void;
    removeNotification: (notification: INotificationItem) => void;
}

export const NotificationContext = React.createContext<INotificationContextProps>({
    notifications: [],
    addNotification: (_notification: INotificationItem) => { },
    removeNotification: (_notification: INotificationItem) => { }
});
