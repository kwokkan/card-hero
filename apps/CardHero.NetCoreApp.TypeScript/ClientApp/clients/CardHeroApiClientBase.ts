// no export as this is used by NSwag generated api clients
class CardHeroApiClientBase {
    getBaseUrl(defaultBaseUrl: string, baseUrl: string): string {
        if (!baseUrl) {
            return defaultBaseUrl;
        }

        if (baseUrl.endsWith('/')) {
            // remove trailing slash as NSwag uses '/' as part of the api url
            return baseUrl.substring(0, baseUrl.length - 2);
        }

        return baseUrl;
    }
}

module.exports = CardHeroApiClientBase;