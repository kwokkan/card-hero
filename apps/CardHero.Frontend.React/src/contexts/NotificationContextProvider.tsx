import React, { useContext, useState } from "react";
import { INotificationItem } from "../types/INotificationItem";
import { NotificationType } from "../types/NotificationType";
import { INotificationContextProps, NotificationContext } from "./NotificationContext";

interface INotificationContextProviderState {
    notifications: INotificationItem[];
}

function NotificationContextProvider(props: any) {
    const [state, setState] = useState<INotificationContextProviderState>({
        notifications: [
            {
                message: "Game will be having maintenance.",
                title: "Status update",
                type: NotificationType.Info
            }
        ]
    });

    const contextProps: INotificationContextProps = {
        addNotification: (notification: INotificationItem) => {
            const newNotifications = [...state.notifications];
            newNotifications.push(notification);

            const newState: INotificationContextProviderState = {
                notifications: newNotifications
            };

            setState(prevState => {
                return { ...prevState, ...newState };
            })
        },
        removeNotification: (notification) => {
            const notifications = [...state.notifications];
            const index = notifications.indexOf(notification);

            if (index > -1) {
                notifications.splice(index, 1);

                const newState: INotificationContextProviderState = {
                    notifications: notifications
                };

                setState(prevState => {
                    return { ...prevState, ...newState };
                })
            }
        },
        notifications: state.notifications
    };

    return (
        <NotificationContext.Provider
            value={contextProps}
            {...props}
        />
    );
}

function useNotificationContext() {
    return useContext(NotificationContext);
}

export { NotificationContextProvider, useNotificationContext };
