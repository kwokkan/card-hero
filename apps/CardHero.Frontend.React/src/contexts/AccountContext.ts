import React from "react";
import { IUserModel } from "../clients/clients";

interface IAccountContextProps {
    user?: IUserModel;
    setUser: (user: IUserModel) => void;
}

export const AccountContext = React.createContext<IAccountContextProps>({ setUser: (_user: IUserModel) => { } });