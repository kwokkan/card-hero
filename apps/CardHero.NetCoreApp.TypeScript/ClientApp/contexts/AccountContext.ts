import React from "react";
import { UserModel } from "../clients/clients";

interface IAccountContextProps {
    user?: UserModel;
    setUser: (user: UserModel) => void;
}

export const AccountContext = React.createContext<IAccountContextProps>({ setUser: (_user: UserModel) => { } });