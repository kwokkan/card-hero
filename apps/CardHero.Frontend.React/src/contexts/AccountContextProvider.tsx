import { ProviderProps, useContext, useState } from "react";
import { IUserModel } from "../clients/clients";
import { AccountContext, IAccountContextProps } from "./AccountContext";

export interface IAccountContextProviderProps {
    user?: IUserModel;
}

interface IAccountContextProviderState {
    user?: IUserModel;
}

function AccountContextProvider(props: ProviderProps<IAccountContextProviderProps>): JSX.Element {
    const [state, setState] = useState<IAccountContextProviderState>({
        user: props.value.user
    });

    const contextProps: IAccountContextProps = {
        setUser: (user) => {
            setState({
                user: user
            });
        },
        user: state.user
    };

    return (
        <AccountContext.Provider
            value={contextProps}
            children={props.children}
        />
    );
}

function useAccountContext() {
    return useContext(AccountContext);
}

function useAccountContextConsumer() {
    return AccountContext.Consumer;
}

export { AccountContextProvider, useAccountContext, useAccountContextConsumer };
