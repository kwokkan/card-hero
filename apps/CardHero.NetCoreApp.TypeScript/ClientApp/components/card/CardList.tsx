import React, { Fragment, useContext } from "react";
import { Link } from "react-router-dom";
import { ICardModel } from "../../clients/clients";
import { AccountContext } from "../../contexts/AccountContext";
import { Icon } from "../../styles/index";
import { getRoutePrefix } from "../../utils/route";

interface ICardListProps {
    cards: ICardModel[];
    routePrefix?: string;
}

export function CardList(props: ICardListProps) {
    const context = useContext(AccountContext);
    const user = context.user;
    const routePrefix = getRoutePrefix(props.routePrefix);

    return (
        <div className="row">
            <table className="table table-striped">
                <thead className="thead-inverse">
                    <tr>
                        <th>Name</th>
                        <th>Health</th>
                        <th>Attack</th>
                        <th>Defence</th>
                        <th>Up Attack</th>
                        <th>Down Attack</th>
                        <th>Left Attack</th>
                        <th>Right Attack</th>
                        <th>Total Stats</th>
                    </tr>
                </thead>

                <tbody>
                    {props.cards.map(c =>
                        <tr key={c.id}>
                            <th scope="row">
                                {user &&
                                    <Fragment>
                                        <Icon icon="star" className={'card-favourite' + (c.isFavourited ? ' enabled' : '')} data-card-id={c.id} />
                                        {' '}
                                    </Fragment>
                                }
                                <Link to={routePrefix + c.id}>{c.name}</Link>
                            </th>
                            <td>{c.health}</td>
                            <td>{c.attack}</td>
                            <td>{c.defence}</td>
                            <td>{c.upAttack}</td>
                            <td>{c.downAttack}</td>
                            <td>{c.leftAttack}</td>
                            <td>{c.rightAttack}</td>
                            <td>{c.totalStats}</td>
                        </tr>
                    )}
                </tbody>
            </table>
        </div>
    );
}
