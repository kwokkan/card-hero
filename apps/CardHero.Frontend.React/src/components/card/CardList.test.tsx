import { MemoryRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import { ICardModel, Rarity, RarityModel } from "../../clients/clients";
import { AccountContext, IAccountContextProps } from "../../contexts/AccountContext";
import { CardList } from "./CardList";

test("<CardList /> renders correctly", () => {
    const cards: ICardModel[] = [
        {
            attack: 1,
            defence: 2,
            description: "First Card desc",
            downAttack: 3,
            health: 4,
            id: 1,
            isFavourited: false,
            leftAttack: 5,
            name: "First Card name",
            rarity: RarityModel.fromJS({
                id: Rarity.Common
            }),
            rightAttack: 6,
            totalStats: 7,
            upAttack: 8,
        },
        {
            attack: 11,
            defence: 12,
            description: "Second Card desc",
            downAttack: 13,
            health: 14,
            id: 2,
            isFavourited: true,
            leftAttack: 15,
            name: "Second Card name",
            rarity: RarityModel.fromJS({
                id: Rarity.Legendary
            }),
            rightAttack: 16,
            totalStats: 17,
            upAttack: 18,
        }
    ];
    const contextValue: IAccountContextProps = {
        user: {
            id: 999
        },
        setUser: () => { }
    };
    const tree = renderer
        .create(
            <AccountContext.Provider value={contextValue}>
                <MemoryRouter>
                    <CardList cards={cards} routePrefix="/jest/" />
                </MemoryRouter>
            </AccountContext.Provider>
        )
        .toJSON();

    expect(tree).toMatchSnapshot();
});
