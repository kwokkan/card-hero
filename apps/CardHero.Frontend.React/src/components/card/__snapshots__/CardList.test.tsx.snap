// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardList /> renders correctly 1`] = `
<div
  className="row"
>
  <table
    className="table table-striped"
  >
    <thead
      className="thead-inverse"
    >
      <tr>
        <th>
          Name
        </th>
        <th>
          Health
        </th>
        <th>
          Attack
        </th>
        <th>
          Defence
        </th>
        <th>
          Up Attack
        </th>
        <th>
          Down Attack
        </th>
        <th>
          Left Attack
        </th>
        <th>
          Right Attack
        </th>
        <th>
          Total Stats
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th
          scope="row"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-star fa-fw card-favourite"
            data-card-id={1}
            data-icon="star"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={{}}
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M381.2 150.3L524.9 171.5C536.8 173.2 546.8 181.6 550.6 193.1C554.4 204.7 551.3 217.3 542.7 225.9L438.5 328.1L463.1 474.7C465.1 486.7 460.2 498.9 450.2 506C440.3 513.1 427.2 514 416.5 508.3L288.1 439.8L159.8 508.3C149 514 135.9 513.1 126 506C116.1 498.9 111.1 486.7 113.2 474.7L137.8 328.1L33.58 225.9C24.97 217.3 21.91 204.7 25.69 193.1C29.46 181.6 39.43 173.2 51.42 171.5L195 150.3L259.4 17.97C264.7 6.954 275.9-.0391 288.1-.0391C300.4-.0391 311.6 6.954 316.9 17.97L381.2 150.3z"
              fill="currentColor"
              style={{}}
            />
          </svg>
           
          <a
            href="/jest/1"
            onClick={[Function]}
          >
            First Card name
          </a>
        </th>
        <td>
          4
        </td>
        <td>
          1
        </td>
        <td>
          2
        </td>
        <td>
          8
        </td>
        <td>
          3
        </td>
        <td>
          5
        </td>
        <td>
          6
        </td>
        <td>
          7
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-star fa-fw card-favourite enabled"
            data-card-id={2}
            data-icon="star"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={{}}
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M381.2 150.3L524.9 171.5C536.8 173.2 546.8 181.6 550.6 193.1C554.4 204.7 551.3 217.3 542.7 225.9L438.5 328.1L463.1 474.7C465.1 486.7 460.2 498.9 450.2 506C440.3 513.1 427.2 514 416.5 508.3L288.1 439.8L159.8 508.3C149 514 135.9 513.1 126 506C116.1 498.9 111.1 486.7 113.2 474.7L137.8 328.1L33.58 225.9C24.97 217.3 21.91 204.7 25.69 193.1C29.46 181.6 39.43 173.2 51.42 171.5L195 150.3L259.4 17.97C264.7 6.954 275.9-.0391 288.1-.0391C300.4-.0391 311.6 6.954 316.9 17.97L381.2 150.3z"
              fill="currentColor"
              style={{}}
            />
          </svg>
           
          <a
            href="/jest/2"
            onClick={[Function]}
          >
            Second Card name
          </a>
        </th>
        <td>
          14
        </td>
        <td>
          11
        </td>
        <td>
          12
        </td>
        <td>
          18
        </td>
        <td>
          13
        </td>
        <td>
          15
        </td>
        <td>
          16
        </td>
        <td>
          17
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
